File: 300.最长递增子序列.py
#
# @lc app=leetcode.cn id=300 lang=python3
#
# [300] 最长递增子序列
#

# @lc code=start
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        dp = [nums[0]]
        length = 1
        for i in range(1, len(nums)):
            if nums[i] > dp[length-1]:
                dp.append(nums[i])
                length = length+1
            else:
                dp[self.findLastSmaller(dp, 0, len(dp)-1, nums[i])] = nums[i]
        return length

    def findLastSmaller(self, dp, left, right, num) -> int:
        loc = right
        while left < right:
            mid = (left+right)//2
            if dp[mid] >= num:
                loc = mid
                right = mid
            else:
                left = mid+1
        return loc
# @lc code=end

File: 300.最长递增子序列.java
import java.util.ArrayList;
import java.util.List;
/*
 * @lc app=leetcode.cn id=300 lang=java
 *
 * [300] 最长递增子序列
 */


// @lc code=start
class Solution {
    public int lengthOfLIS(int[] nums) {
        List<Integer> dp = new ArrayList<Integer>();
        dp.add(nums[0]);
        int length = 1;
        for(int i=1; i<nums.length; i++){
            if(nums[i] > dp.get(length-1)){
                length += 1;
                dp.add(nums[i]);
            }else{
                if(dp.contains(nums[i])){
                    continue;
                } 
                else{
                    dp.set(getFirstBig(dp, nums[i], 0, length-1), nums[i]);
                    
                }
            }
        }
        return length;
    }

    private int getFirstBig(List<Integer> nums, int num, int begin, int end){
        int pos = end;
        while(begin<end){
            int mid = (begin+end)/2;
            if(nums.get(mid)>num){
                end = mid;
                pos = end;
            }else{
                begin = mid+1;
            }
        }
        return pos;
    }
}
// @lc code=end

